Migrate modules to RESTapi

that1guy15 opened the pull request:
Current modules are limited to pyez. I am starting to migrate modules and utils over to using RESTapi. 

Old modules have been moved to 'old_modules' and will be pulled out as we migrate them. 

First up I have converted the following:
- module_util.aos
- library.aos_login
- library.aos_asn_pool

that1guy15 committed:
Migrate modules and utils from AOS PYEZ to RESTapi

pylint cleanup

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,333 @@
+        ranges.append({'first': asn_range[0],
+                       'last': asn_range[1]})
+
+    return ranges
```
can be replaced with one liner (less code is better code):
```python
return [{"first": r[0], "last": r[1]}  for r in pool]
```

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,333 @@
+ENDPOINT = 'resources/asn-pools'
+
+
+def check_ranges_are_valid(module, ranges):
```
Super minor: In case this is not used in playbooks yet, you may consider renaming to `validate_ranges` (looks better to me, but that is very subjective, so completely up to you)

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,333 @@
+    :return: bool
+    """
+    i = 1
+    for asn_range in ranges:
```
use `for i, asn_range in enumerate(ranges, 1):` instead

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,333 @@
+    """
+    margs = module.params
+
+    # If the resource does not exist, return directly
```
unnecessary comment, please remove

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,333 @@
+                         value={})
+
+    # Check if resource is currently in Use or Not
+    if my_pool['status'] != 'not_in_use':
```
double inversions are more difficult to read. Did you mean `if my_pool['status'] == 'in_use'`?

that1guy15 replied:
Need to match any value outside `not_in_use`.

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,209 @@
+    Find an existing resource based on name or id from a a given endpoint
+    :param session: dict
+    :param endpoint: string
+    :param item_name: string
```
documentation says `item_name` and `item_id` are strings, but default values are boolean. Argument types should usually be consistent

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,209 @@
+    data = aos_get(session, endpoint)
+
+    if item_name:
+        my_dict = next((item for item in data['items']
```
Maybe simply:
```python
def find_resource_item(session, endpoint, item_name=None, item_id=None):

    def find_element(data, key):
        for item in data["items"]:
            if item["id"] == key or item["display_name"] == key:
                 return item
         return None

     data = aos_get(session, endpoint)
     key = item_name or item_id or ""

     return find_element(data, key) or {}
```

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,333 @@
+    if not module.check_mode:
+        try:
+            aos_delete(session, ENDPOINT, my_pool['id'])
+        except Exception as e:
```
broad exception

roman-verchikov replied:
can be removed

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,333 @@
+        if not module.check_mode:
+            try:
+                resp = aos_post(session, ENDPOINT, new_pool)
+            except Exception as e:
```
broad exception

roman-verchikov replied:
can be removed

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,209 @@
+                                     headers=headers,
+                                     verify=False)
+
+            if response.status_code == 202:
```
This comment is applicable to all response handling

Any HTTP 2xx should be considered successful.

Replace with:
```
if response.ok:
```

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,209 @@
+
+
+def find_resource_item(session, endpoint,
+                       item_name=False,
```
for consistency purposes with documentation:
```
item_name=None
item_id=None
```

roman-verchikov replied:
```python
def find_resource_by_name(session, endpoint, name):
   ....

def find_resource_by_id(session, endpoint, resource_id):
   ....

def find_resource(session, endpoint, resource_name, resource_id)
    if resource_name:
          rerturn find_resource_by_name(session, endpoint, resource_name)
     elif resource_id:
          return find_resource_by_id(session, endpoint, resource_name)
     return {}
```

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,333 @@
+        my_pool = find_resource_item(margs['session'], ENDPOINT,
+                                     item_name=item_name,
+                                     item_id=item_id)
+    except Exception as e:
```
broad exception

roman-verchikov replied:
can be removed

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,209 @@
+        my_dict = next((item for item in data['items']
+                        if item['id'] == item_id), None)
+
+    if my_dict:
```
```python
return my_dict
```

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,333 @@
+    margs = module.params
+
+    item_name = False
+    item_id = False
```
```
item_name = margs.get('name')
item_id = margs.get('id')
```

roman-verchikov commented on library/aos_login.py:
```diff
@@ -1,108 +1,137 @@
-    HAS_AOS_PYEZ = True
-except ImportError:
-    HAS_AOS_PYEZ = False
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
```
???

roman-verchikov replied:
this should be removed

roman-verchikov commented on library/aos_login.py:
```diff
@@ -1,108 +1,137 @@
+
+    for attempt in range(MAX_ATTEMPTS):
+        try:
+            response = requests.post(aos_url, data=json.dumps(payload), headers=headers, verify=False)
```
Should allow user to specify variable (`verify_certificates=True/False`) instead of using hardcoded value which is insecure.

roman-verchikov commented on library/aos_login.py:
```diff
@@ -1,108 +1,137 @@
+            if response.status_code == 201:
+                aos_session = {"server": mod_args['server'],
+                               "token": response.json()['token']}
+            elif response.status_code == 401:
```
```
else:
    module.fail_json(msg=str(response))
```

roman-verchikov replied:
```
$ curl  -H "Content-Type: application/json"  -k -X POST https://172.20.33.3/api/aaa/login -d '{"username": "admin", "password": "asdf"}'
{"errors": "Invalid credentials"}
```

roman-verchikov commented on library/old_modules/aos_blueprint.py:
```diff
@@ -0,0 +1,293 @@
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.aos import get_aos_session, find_collection_item
```
move old `aos` module to `old_modules/aos.py`:
```python
from ansible.module_utils.old_modules.aos import get_aos_session, find_collection_item
```

roman-verchikov replied:
use `pyez_compat/aos` name instead of `old_modules`?

roman-verchikov commented on library/old_modules/aos_blueprint.py:
```diff
@@ -0,0 +1,293 @@
+        blueprint = aos.Blueprints[name]
+        blueprint.create(template_id, reference_arch=margs['reference_arch'])
+
+    except:
```
broad exception

roman-verchikov commented on library/old_modules/aos_blueprint.py:
```diff
@@ -0,0 +1,293 @@
+        if not module.check_mode:
+            try:
+                blueprint.delete()
+            except:
```
broad exception

roman-verchikov commented on library/old_modules/aos_blueprint.py:
```diff
@@ -0,0 +1,293 @@
+
+    try:
+        aos = get_aos_session(module, margs['session'])
+    except:
```
broad exception

roman-verchikov commented on library/old_modules/aos_blueprint.py:
```diff
@@ -0,0 +1,293 @@
+        my_blueprint = find_collection_item(aos.Blueprints,
+                                        item_name=item_name,
+                                        item_id=item_id)
+    except:
```
broad exception

roman-verchikov commented on library/old_modules/aos_blueprint_param.py:
```diff
@@ -0,0 +1,376 @@
+from ansible.module_utils.pycompat24 import get_exception
+
+try:
+    import yaml
```
?

roman-verchikov replied:
remove try/except and just `import yaml`

roman-verchikov commented on library/old_modules/aos_blueprint_param.py:
```diff
@@ -0,0 +1,376 @@
+try:
+    from apstra.aosom.collection_mapper import CollectionMapper, MultiCollectionMapper
+    HAS_AOS_PYEZ_MAPPER = True
+except ImportError:
```
same here, just `import <>`

roman-verchikov commented on library/old_modules/aos_blueprint_param.py:
```diff
@@ -0,0 +1,376 @@
+
+        try:
+            param_map = yaml.load(param_map_json)
+        except:
```
broad exception

roman-verchikov commented on library/old_modules/aos_blueprint_param.py:
```diff
@@ -0,0 +1,376 @@
+        if not module.check_mode:
+            try:
+                param.value = param_value
+            except:
```
broad exception

roman-verchikov replied:
remove? not clear why this would raise anything

roman-verchikov replied:
```
class Klass():
    @property.setter
     def value(argument):
           send_rocket_to_space()
```

roman-verchikov commented on library/old_modules/aos_blueprint_param.py:
```diff
@@ -0,0 +1,376 @@
+        if not module.check_mode:
+            try:
+                param.value = {}
+            except:
```
same as above

roman-verchikov commented on library/old_modules/aos_blueprint_param.py:
```diff
@@ -0,0 +1,376 @@
+    # --------------------------------------------------------------------
+    try:
+        aos = get_aos_session(module, margs['session'])
+    except:
```
broad exception

roman-verchikov commented on library/old_modules/aos_blueprint_param.py:
```diff
@@ -0,0 +1,376 @@
+        blueprint = find_collection_item(aos.Blueprints,
+                                        item_name=margs['blueprint'],
+                                        item_id=margs['blueprint'])
+    except:
```
broad exception

roman-verchikov commented on library/old_modules/aos_blueprint_param.py:
```diff
@@ -0,0 +1,376 @@
+
+    # aos-pyez availability has been verify already by "check_aos_version"
+    # but this module requires few more object
+    if not HAS_AOS_PYEZ_MAPPER:
```
remove?

roman-verchikov commented on library/old_modules/aos_external_router.py:
```diff
@@ -0,0 +1,329 @@
+    if not module.check_mode:
+        try:
+            # Add Sleep before delete to workaround a bug in AOS
+            time.sleep(2)
```
sleep??

roman-verchikov commented on library/old_modules/aos_external_router.py:
```diff
@@ -0,0 +1,329 @@
+    margs = module.params
+
+    # If the module do not exist, return directly
+    if my_ext_router.exists is False:
```
if not exists?

roman-verchikov commented on library/old_modules/aos_external_router.py:
```diff
@@ -0,0 +1,329 @@
+    # If not in check mode, delete External Router
+    if not module.check_mode:
+        try:
+            # Add Sleep before delete to workaround a bug in AOS
```
which bug???

roman-verchikov commented on library/old_modules/aos_external_router.py:
```diff
@@ -0,0 +1,329 @@
+            # Add Sleep before delete to workaround a bug in AOS
+            time.sleep(2)
+            my_ext_router.delete()
+        except:
```
broad except

that1guy15 replied:
No touching old modules until full refresh.
same for all comments for similar

roman-verchikov commented on library/old_modules/aos_external_router.py:
```diff
@@ -0,0 +1,329 @@
+        my_ext_router = find_collection_item(aos.ExternalRouters,
+                            item_name=item_name,
+                            item_id=item_id)
+    except:
```
broad exception

roman-verchikov commented on library/old_modules/aos_external_router.py:
```diff
@@ -0,0 +1,329 @@
+        module.fail_json(msg="Unable to login to the AOS server")
+
+    item_name = False
+    item_id = False
```
change to `None``

roman-verchikov commented on library/old_modules/aos_hcl.py:
```diff
@@ -0,0 +1,253 @@
+    if not module.check_mode:
+        try:
+            # Need to way 1sec before a delete to workaround a current limitation in AOS
+            time.sleep(1)
```
sleep

roman-verchikov replied:
maybe need to do several repeats instead?

roman-verchikov commented on library/old_modules/aos_template.py:
```diff
@@ -0,0 +1,265 @@
+    if not module.check_mode:
+        try:
+            # need to way 1sec before delete to workaround a current limitation in AOS
+            time.sleep(1)
```
repeat instead?

roman-verchikov reviewed (changes requested):
please address comments

that1guy15 committed:
Added aos_blueprint_deploy and testing

that1guy15 committed:
tox and TravisCI testing setup

roman-verchikov commented on .travis.yml:
```diff
@@ -0,0 +1,5 @@
+language: python
+python:
+  - 3.6
```
Why 3.6? I'd suggest to use the latest version available, which is 3.7

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,300 @@
+        elif asn_range[1] <= asn_range[0]:
+            module.fail_json(msg="2nd element of asn_range {} must be "
+                                 "bigger than 1st ".format(i))
+        i += 1
```
That's not needed. Remove

roman-verchikov commented on library/aos_asn_pool.py:
```diff
@@ -0,0 +1,300 @@
+                        "display_name": my_pool['display_name']}
+
+            for asn_range in my_pool['ranges']:
+                new_pool['ranges'].append({'first': asn_range['first'],
```
this will crash with `AttributeError: 'dict' object has no attribute 'append'`

that1guy15 replied:
`ranges` is a list of dictionaries. Test playbooks update pool successfully.
`{
  "status": "not_in_use",
  "ranges": [
    {
      "status": "pool_element_available",
      "last": 0,
      "first": 0
    }
  ],
  "display_name": "string",
  "tags": [
    "string"
  ]
}`

roman-verchikov replied:
ah, my bad! read it as `new_pool.append()`

roman-verchikov commented on library/pyez_compat/aos.py:
```diff
@@ -0,0 +1,209 @@
+    return resp_data
+
+
+def find_resource_item(session, endpoint,
```
this should be refactored and split into two functions, one `find_resource_by_name` and `find_resource_by_id`.

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,235 @@
+MAX_ATTEMPTS = 3
+
+
+def get_blueprint_status(session, blueprint_id):
```
Every function in this module would send HTTP request 3 times if each request succeeds. `for` loop should be stopped on success.
Also, the code that repeats execution should be isolated in a separate function and reused. Perhaps something as described in this article: https://www.peterbe.com/plog/best-practice-with-retries-with-requests

roman-verchikov replied:
same applies to aos.py

that1guy15 replied:
adding proper breaks. The retry function in the link seems way overkill for these modules as Ansible also has a retry option at the task level.

roman-verchikov replied:
It's not an overkill. You need to isolate the code that is repeated in so many places. This follows DRY sw development principle https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

roman-verchikov replied:
I guess logic with retries could be removed altogether and instead user could use built-in ansible `retries` and `until` feature

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,235 @@
+import urllib3
+from ansible.module_utils.basic import *
+
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
```
Explicit `disable_warnings()` has to be removed from here and everywhere else.

Warnings like these must not be disabled as they holler about the actual security threat
so they could be disabled in development environment by temporarily setting env variable like `export PYTHONWARNINGS="ignore:Unverified HTTPS request"`, but should be present when customer uses this module.

roman-verchikov replied:
this still needs to be resolved

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,235 @@
+                requests.Timeout) as e:
+            return "Unable to connect to server {}: {}".format(aos_url, e)
+
+    bp_id = [i['id'] for i in resp_data['items']
```
`s/bp_id/bp_ids/g`

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,235 @@
+                requests.Timeout) as e:
+            return "Unable to connect to server {}: {}".format(aos_url, e)
+
+    bp_ver = int(resp_data['version'])
```
this may potentially raise TypeError or ValueError in case `int()` fails to convert provided value to integer. What stops us from simply returning the `resp_data['version']`?

roman-verchikov replied:
unresolved

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,235 @@
+    for attempt in range(MAX_ATTEMPTS):
+        try:
+            response = requests.put(aos_url,
+                                    data=json.dumps(payload),
```
`json=payload`

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,235 @@
+            response = requests.put(aos_url,
+                                    data=json.dumps(payload),
+                                    headers=headers,
+                                    verify=False)
```
certificates must be verified by default, but user should be able to change behavior in playbook.

roman-verchikov commented on library/pyez_compat/aos.py:
```diff
@@ -0,0 +1,209 @@
+        try:
+            response = requests.get(aos_url, headers=headers, verify=False)
+
+            if response.status_code == 200:
```
201 may also be a successful return, as well as any other 2xx HTTP code. This condition is too restrictive and should be changed

roman-verchikov commented on library/pyez_compat/aos.py:
```diff
@@ -0,0 +1,209 @@
+                                     headers=headers,
+                                     verify=False)
+
+            if response.status_code == 202:
```
same as above

roman-verchikov commented on library/pyez_compat/aos.py:
```diff
@@ -0,0 +1,209 @@
+                                    headers=headers,
+                                    verify=False)
+
+            if response.status_code == 202:
```
ditto

roman-verchikov commented on library/pyez_compat/aos.py:
```diff
@@ -0,0 +1,209 @@
+        try:
+            response = requests.delete(aos_url, headers=headers, verify=False)
+
+            if response.status_code == 202:
```
ditto

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+def check_ranges_are_valid(module, ranges):
+
+    i = 1
+    for range in ranges:
```
enumerate

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+def get_list_of_range(asn_pool):
+    ranges = []
+
+    for range in asn_pool.value['ranges']:
```
`range` is a python language keyword, rename to `r` or something

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+
+    # Create value
+    datum = dict(display_name=name, ranges=[])
+    for range in ranges:
```
same as above

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+
+    asn_pool.datum = datum
+
+    ## Write to AOS
```
unnecessary comment. remove

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+
+def create_new_asn_pool(asn_pool, name, ranges):
+
+    # Create value
```
unnecessary comment

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+    if not module.check_mode:
+        try:
+            my_pool.delete()
+        except:
```
broad exception

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+    margs = module.params
+
+    # If the module do not exist, return directly
+    if my_pool.exists is False:
```
simply `if not my_pool.exists:`

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+            try:
+                my_new_pool = create_new_asn_pool(my_pool, margs['name'], margs['ranges'])
+                my_pool = my_new_pool
+            except:
```
broad exception

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+            module.fail_json(msg="Unable to find display_name in 'content', Mandatory")
+
+    # if asn_pool doesn't exist already, create a new one
+    if my_pool.exists is False and 'name' not in margs.keys():
```
simply `if 'name' not in margs:`

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+
+        if not module.check_mode:
+            try:
+                my_new_pool = create_new_asn_pool(my_pool, margs['name'], margs['ranges'])
```
potential race condition: asn pool may be created from elsewhere right after `my_poool.exists` returns `False`

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+        my_pool = find_collection_item(aos.AsnPools,
+                            item_name=item_name,
+                            item_id=item_id)
+    except:
```
broad exception

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+
+    try:
+        aos = get_aos_session(module, margs['session'])
+    except:
```
broad exception handling which hides reason for failure and does not provide information on why the login actually failed (failed to establish connection, wrong credentials, broken code?)

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+                            item_name=item_name,
+                            item_id=item_id)
+    except:
+        module.fail_json(msg="Unable to find the IP Pool based on name or ID, something went wrong")
```
extremely unhelpful message, catch specific exceptions and display error to a user

roman-verchikov commented on library/pyez_compat/aos_asn_pool.py:
```diff
@@ -0,0 +1,335 @@
+    )
+
+    # Check if aos-pyez is present and match the minimum version
+    check_aos_version(module, '0.6.0')
```
version must be verified by requirements, not in runtime

that1guy15 committed:
remove old pyez modules and utils

Cleanup from comments

gitignore unwanted .idea

correct travis for python 3.7

better exception handling with aos.py

better exception handling with aos.py

resolve ssl verify disabled.

clean up tox

clean up from comments

clean up from comments

clean up from comments

clean up from comments

clean up from comments

roman-verchikov commented on tox.ini:
```diff
@@ -1,36 +1,17 @@
-# E501: 80 character line length limit
-# E128: continuation line under-indented for visual indent
-ignore = E501, E128
+select = E501
```
Remove that. This only runs a single test for the code which only checks the line length.

roman-verchikov commented on tox.ini:
```diff
@@ -1,36 +1,17 @@
-deps=sphinx
-commands=
-    sphinx-build -W -b html -d {envtmpdir}/doctrees ./source {envtmpdir}/html
+commands = nosetests -v
```
consider using [pytest](https://pytest.readthedocs.io/) instead

roman-verchikov commented on requirements.txt:
```diff
@@ -0,0 +1,8 @@
+ansible>=2.7.9
+flake8<=3.7.7
+mock>=2.0.0
+nose>=1.3.7
```
`flake8`, `nose`, `mock`, `pylint` and `tox` should be moved to test-requirements.txt

roman-verchikov replied:
make sure test-related dependencies are only present in `test-requirements.txt` and not in `requirements.txt`

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,225 @@
+        except (requests.ConnectionError,
+                requests.HTTPError,
+                requests.Timeout) as e:
+            return "Unable to connect to server {}: {}".format(aos_url, e)
```
Instead of returning you should retry

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,225 @@
+            response = requests.post(aos_url,
+                                     data=json.dumps(payload),
+                                     headers=headers,
+                                     verify=False)
```
Certificates must be verified by default. Also module user must be allowed to pass certificate as argument

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,229 @@
+
+            if response.ok:
+                return response.json()
+                break
```
break after return

roman-verchikov replied:
also see my comment above

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,229 @@
+        return {}
+
+
+def find_resource_by_id(resource_data, id):
```
`id` is a built-in python function name (https://docs.python.org/2/library/functions.html#id). Rename the parameter to something else, e.g. `resource_id`

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,229 @@
+    return resp_data
+
+
+def find_resource_by_name(resource_data, name):
```
these two functions can be further isolated:

```python
def _find_resource(resource_data, key, keyword):
     for item in resource_data['items']:
          if item[keyword] == key:
               return item
     return {}

def find_resource_by_name(resource_data, name):
     return _find_resource(resource_data, name, "display_name")

def find_resource_by_id(resource_data, resource_id):
     return _find_resource(resource_data, resource_id, "id")
```

that1guy15 replied:
Yes, they originally where, but we split them out to reduce if/else login and simplify. half-dozen, 6 of the other.

roman-verchikov replied:
I guess you misunderstood my comment. I meant `find_resource_by_name`  and `find_resource_by_id` still have common code that can be isolated in another function. And I also gave an example of how it can be implemented.

that1guy15 replied:
Yup misunderstood and for some reason missed the code snippet when I read this.

roman-verchikov commented on test-requirements.txt:
```diff
@@ -0,0 +1,5 @@
+mock>=2.0.0
+nose>=1.3.7
+pylint>=2.3.1
+tox>=3.7.0
```
it's usually better to use strict version matching (using `==` operand) to avoid potentially broken dependencies if one or more of them updates

roman-verchikov replied:
this also applies to `requirements.txt`

roman-verchikov commented on tests/test_aos.py:
```diff
@@ -0,0 +1,92 @@
+}
+
+
+class test_find_resource(unittest.TestCase):
```
class name violates pep8: should be in CamelCase.

roman-verchikov commented on tox.ini:
```diff
@@ -1,36 +1,18 @@
-ignore = E501, E128
+max-line-length = 85
+exclude = library/pyez_compat
+ignore = E402,F403,F405,
```
Fix code instead of making flake8 ignore some errors. flake8 should report **all** violations of pep8, as well as all code in the module must follow pep8 standard.

that1guy15 replied:
These all come from Ansible format for modules and imports.

roman-verchikov replied:
Nope, flake8 doesn't scan dependencies. All the errors look valid to me and should be resolved.  This is the output I got after removing `ignore = ...` line from `tox.ini`

```
$ tox -e flake8
flake8 installed: DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.,configparser==3.7.4,entrypoints==0.3,enum34==1.1.6,flake8==3.7.7,functools32==3.2.3.post2,mccabe==0.6.1,pycodestyle==2.5.0,pyflakes==2.1.1,typing==3.6.6
flake8 run-test-pre: PYTHONHASHSEED='2868781585'
flake8 runtests: commands[0] | flake8 library module_utils tests
library/aos_login.py:81:1: E402 module level import not at top of file
library/aos_login.py:81:1: F403 'from ansible.module_utils.basic import *' used; unable to detect undefined names
library/aos_login.py:82:1: E402 module level import not at top of file
library/aos_login.py:98:43: F405 'json' may be undefined, or defined from star imports: ansible.module_utils.basic
library/aos_login.py:112:14: F405 'AnsibleModule' may be undefined, or defined from star imports: ansible.module_utils.basic
library/aos_asn_pool.py:115:1: E402 module level import not at top of file
library/aos_asn_pool.py:116:1: E402 module level import not at top of file
library/aos_blueprint_deploy.py:71:1: E402 module level import not at top of file
library/aos_blueprint_deploy.py:71:1: F403 'from ansible.module_utils.basic import *' used; unable to detect undefined names
library/aos_blueprint_deploy.py:72:1: E402 module level import not at top of file
library/aos_blueprint_deploy.py:149:42: F405 'json' may be undefined, or defined from star imports: ansible.module_utils.basic
library/aos_blueprint_deploy.py:172:14: F405 'AnsibleModule' may be undefined, or defined from star imports: ansible.module_utils.basic
```

The reason those are present is because there's a number of places where `from ansible.module_utils.basic import *` which is [considered bad practice](https://www.python.org/dev/peps/pep-0008/#imports):

> Wildcard imports (from <module> import *) should be avoided, as they make it unclear which names are present in the namespace, confusing both readers and many automated tools.

roman-verchikov replied:
Ok, but that only suggest to ignore E402, not 403 and 405. And also regarding wildcard imports - those are [not allowed anymore](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#python-imports). The link you provided is for ansible 2.3, while in requirements.txt we specify `ansible>=2.7.9`.

Bottom line: `tox.ini` should only have `ignore = E402`, the rest of flake8 issues should be resolved.

that1guy15 commented:
Actually the wildcard import was a requirement of ansible but looks like it
has been removed.
https://docs.ansible.com/ansible/2.3/dev_guide/developing_modules_general.html#common-module-boilerplate

For import not at top of file, Ansible expects the module docs and
description to be at top of file so they can be read. This is how Ansible
wants modules formated.
https://docs.ansible.com/ansible/2.3/dev_guide/developing_modules_documenting.html

On Thu, Apr 25, 2019 at 5:58 PM Roman Verchikov <notifications@github.com>
wrote:

> *@roman-verchikov* commented on this pull request.
> ------------------------------
>
> In tox.ini
> <https://github.com/Apstra/aos-ansible/pull/14#discussion_r278761224>:
>
> >  [flake8]
> -# Ignore the following pep8 violations
> -# E501: 80 character line length limit
> -# E128: continuation line under-indented for visual indent
> -ignore = E501, E128
> +max-line-length = 85
> +exclude = library/pyez_compat
> +ignore = E402,F403,F405,
>
> Nope, flake8 doesn't scan dependencies. All the errors look valid to me
> and should be resolved. This is the output I got after removing ignore =
> ... line from tox.ini
>
> $ tox -e flake8
> flake8 installed: DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.,configparser==3.7.4,entrypoints==0.3,enum34==1.1.6,flake8==3.7.7,functools32==3.2.3.post2,mccabe==0.6.1,pycodestyle==2.5.0,pyflakes==2.1.1,typing==3.6.6
> flake8 run-test-pre: PYTHONHASHSEED='2868781585'
> flake8 runtests: commands[0] | flake8 library module_utils tests
> library/aos_login.py:81:1: E402 module level import not at top of file
> library/aos_login.py:81:1: F403 'from ansible.module_utils.basic import *' used; unable to detect undefined names
> library/aos_login.py:82:1: E402 module level import not at top of file
> library/aos_login.py:98:43: F405 'json' may be undefined, or defined from star imports: ansible.module_utils.basic
> library/aos_login.py:112:14: F405 'AnsibleModule' may be undefined, or defined from star imports: ansible.module_utils.basic
> library/aos_asn_pool.py:115:1: E402 module level import not at top of file
> library/aos_asn_pool.py:116:1: E402 module level import not at top of file
> library/aos_blueprint_deploy.py:71:1: E402 module level import not at top of file
> library/aos_blueprint_deploy.py:71:1: F403 'from ansible.module_utils.basic import *' used; unable to detect undefined names
> library/aos_blueprint_deploy.py:72:1: E402 module level import not at top of file
> library/aos_blueprint_deploy.py:149:42: F405 'json' may be undefined, or defined from star imports: ansible.module_utils.basic
> library/aos_blueprint_deploy.py:172:14: F405 'AnsibleModule' may be undefined, or defined from star imports: ansible.module_utils.basic
>
> The reason those are present is because there's a number of places where from
> ansible.module_utils.basic import * which is considered bad practice
> <https://www.python.org/dev/peps/pep-0008/#imports>:
>
> Wildcard imports (from import *) should be avoided, as they make it
> unclear which names are present in the namespace, confusing both readers
> and many automated tools.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Apstra/aos-ansible/pull/14#discussion_r278761224>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARQHKTMTMGZ65RQTDV3QZDPSIZQ5ANCNFSM4HDUVGUA>
> .
>


-- 
*Ryan Booth *
Phone: 806.679.0903
email: boothr2@gmail.com

roman-verchikov commented on tox.ini:
```diff
@@ -1,36 +1,19 @@
-ignore = E501, E128
+max-line-length = 85
+exclude = library/pyez_compat
+ignore = E402
```
would also be good to add a comment and a link pointing to the Ansible documentation

roman-verchikov commented on requirements.txt:
```diff
@@ -1 +1,8 @@
+mock==2.0.0
+nose==1.3.7
+pylint==2.3.1
+requests==2.21.0
```
flake8, mock, nose and pylint should only be present in test-requirements.txt

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,196 @@
+from requests.adapters import HTTPAdapter
+
+
+def set_requests_verify():
```
This doesn't look right. `PYTHONWARNINGS` environment variable should not impact validation of certificates, it is only responsible for printing or not printing warnings in certain modules. This additional behavior looks confusing. It would be ok to introduce another environment variable `AOS_SSL_CERT_VERIFICATION`, and control `ssl` argument of `requests.<method>(ssl=<>)`, for example:

```python
session.request(method, url, headers, ssl=os.environ.get("AOS_SSL_CERT_VERIFICATION", True))
```

that1guy15 replied:
Im not sure what this does and cant find anything on `ssl` argument. This example is not working either. Ive changed the environment variable to reduce confusion.

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,188 @@
+
+    else:
+        module.fail_json(
+            msg="Issue deploying blueprint {}: {}"
```
AOS typically returns JSON with a user-friendly error message, e.g.
```
curl -X POST "https://172.20.100.3/api/aaa/login" -H "accept: application/json" -H "content-type: application/json" -d "{ \"username\": \"asdf\", \"password\": \"zxcv\"}"
{
  "errors": "Invalid credentials"
}
```

Instead in this case error message would contain not very helpful string `<Response [401]>`:

```python
>>> resp = requests.post("https://172.20.100.3/api/aaa/login", json={"username": "asdf", "password": "zxcv"}, verify=False)
>>> print(resp)
<Response [401]>
```

Would be better with

```python
error_message = str(response)
try:
     error_message = response.json().get('errors')
except (TypeError, ValueError) as e:
     # attempt to decode error message from response
     # and ignore cases when JSON is not available
     LOG.info("Failed to decode JSON from response: %s, error: %s", response, e)
msg="Issue deploying blueprint {}: {}".format(bp_id, error_message))
```

that1guy15 replied:
My idea is to return simple errors here not full output of why AOS will not deploy. That is left to the user to further define in their playbook if they want.

roman-verchikov replied:
Well that would be problematic for a user to define anything in their playbook, because there will be very little information in error message as I mentioned. Most information about error is stored in JSON response from AOS server, while existing code only returns error message based on HTTP error code (e.g. "404 Not found", or "409 - Conflict"), without important details that are provided in JSON response body.

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,188 @@
+    aos_url = "https://{}/api/blueprints/{}/deploy" \
+        .format(mod_args['session']['server'], bp_id)
+
+    headers = {'AUTHTOKEN': mod_args['session']['token'],
```
AUTHTOKEN is `session['token']` everywhere else, but here it's `mod_args['session']['token']`. Not sure why it's done this way but looks suspicious.

Also, headers are pretty much identical in every single call, would be good to isolate them

that1guy15 replied:
needed lots of rework, current state was built for existing project as a self-containing module w/o `module_utils.aos.py`

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,195 @@
+
+
+def set_requests_verify():
+    if 'AOS_VALID_CERT' in os.environ and os.environ['AOS_VALID_CERT'] == 'false':
```
Variable naming could be better... I would assume `AOS_VALID_CERT` is expected to have only boolean values? In this case it would be better to name it `AOS_VERIFY_CERTIFICATE` or even `AOS_ENABLE_CERT_VERIFICATION`. Allowed values should be in the set of `{'True', 'true', 'TRUE', 1, 'YES', 'Yes', 'yes', 'y', 'Y', 'False', 'FALSE', 'false', 0, 'No', 'NO', 'no', 'n', 'N'}`

roman-verchikov commented on tox.ini:
```diff
@@ -1,36 +1,21 @@
+max-line-length = 85
+exclude = library/pyez_compat
+ignore = E402
+# Ansible docs required at top of file
```
usually comments are placed above the line, in that case I would recommend moving `ignore = E402` below the comment

roman-verchikov commented on tox.ini:
```diff
@@ -1,36 +1,21 @@
+exclude = library/pyez_compat
+ignore = E402
+# Ansible docs required at top of file
+# https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html
```
Please add a newline, github places that nasty red symbol suggesting newline in the end of the file

that1guy15 commented:
@roman-verchikov any last comments?

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,172 @@
+
+    resp_data = aos_get(session, endpoint)
+
+    return int(resp_data['version'])
```
Minor, but I'm still not sure as to why conversion to `int()` is required here, it may potentially raise `TypeError` or `ValueError` exceptions in case conversion fails.

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,172 @@
+
+    resp_data = aos_get(session, endpoint)
+
+    bp_id = [i['id'] for i in resp_data['items']
```
Current implementation of `get_blueprint_id` returns different types of values: string with blueprint ID if blueprint was found, or empty list if blueprint was not found. Return value types have to be consistent. I suggest returning `None` in case blueprint was not found:

```python
for i in resp_data['items']:
    if i['label'] == blueprint_name:
        return i['id']
return None
```

roman-verchikov commented on library/aos_blueprint_deploy.py:
```diff
@@ -0,0 +1,172 @@
+        except (TypeError, ValueError) as e:
+            module.fail_json(
+                msg="Failed to decode JSON from response: {}, error: {}"
+                    .format(response, e))
```
this code might be useful in other places too, consider moving to shared module.

that1guy15 replied:
Ill look at this with my next round of modules Im working on.

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,199 @@
+def set_requests_verify():
+
+    if 'AOS_VERIFY_CERTIFICATE' in os.environ and \
+            os.environ['AOS_VERIFY_CERTIFICATE'] in valid_bool_false:
```
could be replaced with:

```python
def set_requests_verify():
    return os.environ.get('AOS_VERIFY_CERTIFICATE', 'Yes') in valid_bool_false
```

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,199 @@
+    else:
+        response.raise_for_status()
+
+    return resp_data
```
last `return` statement is never executed, can be simplified to

```python
def requests_response(response):
    return response.json() if response.ok else response.raise_for_status()
```

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,199 @@
+    headers = {'AUTHTOKEN': session['token'],
+               'Accept': "application/json",
+               'Content-Type': "application/json",
+               'cache-control': "no-cache"}
```
`headers` are repeated in all functions of this module. Would be good to isolate in a function and reuse

roman-verchikov commented on tests/test_aos.py:
```diff
@@ -0,0 +1,92 @@
+}
+
+
+class testFindResource(unittest.TestCase):
```
Class name [should start with capital letter](https://www.python.org/dev/peps/pep-0008/#class-names)

roman-verchikov commented on tests/test_aos_login.py:
```diff
@@ -0,0 +1,66 @@
+    raise AnsibleFailJson(kwargs)
+
+
+def module_exit_mock(mocker):
```
this function doesn't seem to be used anywhere, please remove if that's the case

roman-verchikov commented on tox.ini:
```diff
@@ -1,36 +1,21 @@
-deps = {[base]deps}
-setenv =
-    _TOPDIR = {toxinidir}
+    -r{toxinidir}/test-requirements.txt
```
I guess `{toxinidir}` is not required here, but that depends on wether `tox` is expected to run from non-root project directory, which is not the case, because other directories are listed without `{toxinidir}`.
```
deps = 
    -rrequirements.txt
    -rtest-requirements.txt
```

roman-verchikov commented on module_utils/aos.py:
```diff
@@ -0,0 +1,199 @@
+valid_bool_false = {'False', 'FALSE', 'false', 0, 'No', 'NO', 'no', 'n', 'N'}
+
+
+def set_requests_verify():
```
use ansible `convert_bool` module instead to maintain full compatibility

```python
from ansible.module_utils.parsing.convert_bool import boolean

def set_requests_verify()
    return boolean(os.environ.get('AOS_VERIFY_CERTIFICATE', 'yes'))
```

roman-verchikov commented on tox.ini:
```diff
@@ -1,36 +1,21 @@
-# E128: continuation line under-indented for visual indent
-ignore = E501, E128
+max-line-length = 85
+exclude = library/pyez_compat
```
this is not needed anymore, `pyez_compat` directory was removed

roman-verchikov reviewed (approved):
One final comment and good to merge

that1guy15 committed:
clean up from comments
